Mender support

dberliner opened the pull request.

dberliner commented on package/go/go.mk:
```diff
@@ -61,20 +61,22 @@ endif
-	GO_ASSUME_CROSSCOMPILING=1
+	CXX=$(HOSTCXX_NOCCACHE) 
 
+ifeq ($(BR2_PACKAGE_HOST_GO_TARGET_ARCH_SUPPORTS),y)
```
To be clear what we're doing, this will produce a compiler that can be run on the target if it is supported. A compiler that can run on the host is always built and that compiler can target all supported systems.

drbild replied:
So there are really three things:

1) Go compiler to run on the host (if so, can build host packages)
2) Go compiler to run on the host and support the target arch (if so, can build target packages)
3) Go compiler to run on the target.

However, I don't think Buildroot should support option 3.  They have an explicit policy of [not adding compilers to the target image](https://buildroot.org/downloads/manual/manual.html#faq-no-compiler-on-target).  So there's no reason to even build the Go compiler for the target arch.

So this can be simplified to just build the Go compiler for the host.  No `GOARCH`, etc.) when building the compiler.

The target support is then just (1) is the arch supported and (2) what `GOARCH`, etc. to set when compiling target packages.

drbild replied:
So for submission upstream, I'd split this into four commits, starting from the current Buildroot master.

1) Do not build the go compiler for the target arch.
1) Rework the `*_SUPPORTS` flags as suggested in Thomas's email.
1) Add the `host-golang-package` command (basically Mirza's patch, but slightly modified).
1) Add the mender artifact package (basically Mirza's patch).

dberliner replied:
I've removed the code accomplishing 3 and 2 has been accomplished by modifying packages that call target builds to use the new variable.

drbild commented on package/go/Config.in.host:
```diff
@@ -2,6 +2,11 @@ config BR2_PACKAGE_HOST_GO_ARCH_SUPPORTS
 	default y
 	depends on BR2_PACKAGE_HOST_GO_BOOTSTRAP_ARCH_SUPPORTS
+
+config BR2_PACKAGE_HOST_GO_TARGET_ARCH_SUPPORTS
```
Are there any existing Go packages that need to be updated to depend on this new variable, instead of the old `BR2_PACKAGE_HOST_GO_ARCH_SUPPORTS`?

dberliner replied:
Not within buildroot, it wouldn't surprise me if the HOST variable is being used as a proxy for target support in some packages but that's not ours to change. We should update the documentation, though.

dberliner replied:
Actually, looking at it further I think they all warrant changing. The condition is now much easier to pass (no longer asking if the target supports go) and could cause build commands to be issued for non-supported packages. There's only a few to update so it won't be difficult .

drbild reviewed (changes requested):
A few comments, mostly related to getting this upstream.

drbild reviewed (commented):
I think this looks good functionally.

To get it accepted upstream, I think the commits will need to be reorganized so that
- each commit is one logical change
- no commit breaks anything.

Currently the commit to spit the `*_SUPPORTS` variables breaks building some packages. That isn't fixed until the last commit that updates those packages.

Also, the commits will need to have a detailed commit message explaining the change.

So I still think you'll want the following series:
1) do not build the Go toolchain to run on the target (one logical change, fixes what is essentially unwanted behavior in the current master)
1) separate Go host and target arch support (this commit will also update any packages that depend on target support)
1) add support to build Go packages for host
1) add mender-artifact package.

Commit:
Do not build Go toolchain for target arch.

GOARCH, GOOS and other variables during build are for creating a
toolchain that can be run on the provided combination. Buildroot does
not allow this feature. The Go toolchain is able to target all
supported targets by default so further variables are not needed.

Commit:
Create seperate variable for target Go support.

Creates a new variable BR2_PACKAGE_HOST_GO_TARGET_ARCH_SUPPORTS to
denotate Go support on the target arch. This was previously done with
BR2_PACKAGE_HOST_GO_ARCH_SUPPORTS which will now denotate host Go
support. Go is written in Go so target support will always require host.
No additional make steps are required becuase Go can cross-compile to
all supoprted targets by default.

A new new make command "host-golang-package" has also been added to
build Go packages on the host. The "golang-package" still supports
target builds like before.

A few packages use BR2_PACKAGE_HOST_GO_ARCH_SUPPORTS for target go
builds so they have been modified to use the new variable.

Commit:
Add mender-artifact package

mender-artifact is a host tool to generate update images
in the Mender artifact file format.

Example usage:

    $ mender-artifact write rootfs-image \
         --update rootfs.ext4 \
         --output-path rootfs.mender \
         --artifact-name "release-v1.0.0" \
         --device-type "beaglebone"

Above will generate a Mender artifact called "rootfs.mender"
containing the "rootfs.ext4" image along with meta-data.

One can read-out the meta-data with the following command:

    $ mender-artifact read rootfs.mender
      Mender artifact:
      Name: release-v1.0.0
      Format: mender
      Version: 2
      Signature: no signature
      Compatible devices: '[beaglebone]'

    Updates:
      0000:
        Type:   rootfs-image
        Files:
          name:     rootfs.ext4
          size:     52428800
          modified: 2018-08-27 09:10:55 +0200 CEST
          checksum: e70b113fb0964a810a3043586eb4fc1c48e684ba78b02ba65fead4aa3e540d87

mirzak commented:
Overall looks great!

Couple of remarks,

- You will need to add Signed-off-by, tags for all commit. That is part of the commit message
- The current release of "mender-artifact" is 2.4.0, we might as well bump that as I noticed that the current package is 2.2.0.

Commit:
Add position independent code flag to fix compiler error

mirzak commented:
It seems that someone else also revived my initial patch series, https://patchwork.ozlabs.org/cover/1037529/

drbild commented:
Nice!  We'll see if that gets accepted.

@dberliner After our current deadline, you could try Angelo Compagnucci's patches instead of ours. If they work, then send a reply with `Tested-by: Daniel Berliner <daniel.berliner@xaptum.com>` for each to the listserv.

dberliner commented:
I do not remember what happened her but this issue was solved a very long time ago.
